app/src/main/java/com/cmput301f20t21/bookfriends/ui/borrow/accepted/AcceptedDetailViewModel.java

```diff
@@ -0,0 +1,109 @@
+import com.google.firebase.firestore.ListenerRegistration;
+
+import java.util.Arrays;
+
+public class AcceptedDetailViewModel extends ViewModel {
+    private RequestRepository requestRepository;
+    private BookRepository bookRepository;
+
+    private final MutableLiveData<Request> request = new MutableLiveData<>();
+    private final MutableLiveData<SCAN_ERROR> errorMessage = new MutableLiveData<>();
+
+    private ListenerRegistration listenerRegistration;
+
+    public AcceptedDetailViewModel() {
+        this(RequestRepositoryImpl.getInstance(), BookRepositoryImpl.getInstance());
+    }
+
+    public AcceptedDetailViewModel(RequestRepository requestRepository, BookRepository bookRepository) {
+        this.requestRepository = requestRepository;
+        this.bookRepository = bookRepository;
+    }
+
+    public LiveData<Request> getRequest(Book book) {
+        fetchRequest(book.getId());
+        return request;
+    }
+
+    public MutableLiveData<SCAN_ERROR> getErrorMessage() {
+        return errorMessage;
+    }
+
+    private void fetchRequest(String bookId) {
+        requestRepository
+                .getRequestsByBookIdAndStatus(bookId, Arrays.asList(REQUEST_STATUS.ACCEPTED, REQUEST_STATUS.HANDING))
+                .addOnSuccessListener(requests -> {
+                    request.setValue(requests.get(0));
+                    registerSnapshotListener();
+                })
+                .addOnFailureListener(e -> errorMessage.setValue(SCAN_ERROR.UNEXPECTED));
+    }
+
+    public void handleScannedIsbn(Book currentBook, String scannedIsbn, OnSuccessCallbackWithMessage<Book> successCallback) {
+        if (currentBook.getIsbn().equals(scannedIsbn)) {
+            requestRepository.updateRequestStatus(request.getValue(), REQUEST_STATUS.BORROWED)
+                    .addOnSuccessListener(updatedRequest -> {
+                        bookRepository.updateBookStatus(currentBook, BOOK_STATUS.BORROWED)
+                                .addOnSuccessListener(updatedBook -> {
+                                    request.setValue(updatedRequest);
+                                    successCallback.run(updatedBook);
+                                })
+                                .addOnFailureListener(e -> errorMessage.setValue(SCAN_ERROR.UNEXPECTED));
+                    })
+                    .addOnFailureListener(e -> errorMessage.setValue(SCAN_ERROR.UNEXPECTED));
+        } else {
+            errorMessage.setValue(SCAN_ERROR.INVALID_ISBN);
+        }
+    }
+
+    // https://stackoverflow.com/questions/48699032/how-to-set-addsnapshotlistener-and-remove-in-populateviewholder-in-recyclerview
+    public void registerSnapshotListener() {
+        if (listenerRegistration == null) {
+            Request currentRequest = request.getValue();
+            if (currentRequest != null) {
+                requestRepository.getRefById(currentRequest.getId()).addSnapshotListener((snapshot, error) -> {
```

Comment:
https://firebase.google.com/docs/firestore/query-data/listen#detach_a_listener
do we need to assign the `addSnapshotListener` return value to listenerRegistration?

Reply:
nice catch, yes we do, i missed it :)